### PR TITLE
[SPARK-12939][SQL] migrate encoder resolution logic to Analyzer

### PR DESCRIPTION
**Pull request by cloud-fan**
https://issues.apache.org/jira/browse/SPARK-12939

Now we will catch `ObjectOperator` in `Analyzer` and resolve the `fromRowExpression/deserializer` inside it.  Also update the `MapGroups` and `CoGroup` to pass in `dataAttributes`, so that we can correctly resolve value deserializer(the `child.output` contains both groupking key and values, which may mess things up if they have same-name attribtues). End-to-end tests are added.

follow-ups:
- remove encoders from typed aggregate expression.
- completely remove resolve/bind in `ExpressionEncoder`

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @marmbrus @yhuai @rxin

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/object.scala [hunk 1]
```diff
@@ -30,6 +30,15 @@ trait ObjectOperator extends LogicalPlan {
   /** The serializer that is used to produce the output of this operator. */
   def serializer: Seq[NamedExpression]
 
+  override def output: Seq[Attribute] = serializer.map(_.toAttribute)
+
+  /**
+   * An [[ObjectOperator]] may have one or more deserializers to convert internal rows to objects.
+   * For each deserializer, [[ObjectOperator]] should also provide the attributes that can resolve
+   * this deserializer.
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/object.scala]: "An [[ObjectOperator]] must also provide the attributes that are available during the resolution of each deserializer."

File: sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala [hunk 2]
```diff
@@ -142,12 +143,14 @@ case class MapGroups(
...
-    leftObject: Expression,
-    rightObject: Expression,
+    keyDeserializer: Expression,
+    leftDeserializer: Expression,
+    rightDeserializer: Expression,
     serializer: Seq[NamedExpression],
     leftGroup: Seq[Attribute],
     rightGroup: Seq[Attribute],
+    leftAttr: Seq[Attribute],
+    rightAttr: Seq[Attribute],
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala]: Why add these arguments when in practice they are always the same as `left.output`/`right.output`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/object.scala [hunk 3]
```diff
@@ -44,18 +53,30 @@ trait ObjectOperator extends LogicalPlan {
...
+ * A dummy [[ObjectOperator]] that is used to trigger deserializer resolution in Analyzer and we
+ * can get back the resolved deserializer later.
+ */
+case class DummyObjectOperator(deserializer: Expression, attributes: Seq[Attribute])
+  extends LeafNode with ObjectOperator {
+
+  override def inputSet: AttributeSet = AttributeSet(attributes)
+  override def serializer: Seq[NamedExpression] = Nil
+  override def deserializers: Seq[(Expression, Seq[Attribute])] = Seq(deserializer -> attributes)
+}
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/object.scala]: Could we instead wrap the resolution logic in a public function that is defined on `ResolveReferences` and just call that directly when we need to?  The flow is pretty hard to follow here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/object.scala [hunk 4]
```diff
@@ -44,18 +53,30 @@ trait ObjectOperator extends LogicalPlan {
   def withObjectOutput: LogicalPlan = if (output.head.dataType.isInstanceOf[ObjectType]) {
     this
   } else {
-    withNewSerializer(outputObject)
+    withNewSerializer(outputObject :: Nil)
   }
 
   /** Returns a copy of this operator with a different serializer. */
-  def withNewSerializer(newSerializer: NamedExpression): LogicalPlan = makeCopy {
+  def withNewSerializer(newSerializer: Seq[NamedExpression]): LogicalPlan = makeCopy {
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/object.scala]: nit: `withNewSerializers`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 5]
```diff
@@ -443,25 +444,32 @@ class Analyzer(
...
-            ExtractValue(child, fieldExpr, resolver)
-        }
+        val newG = resolveExpression(generator, child)
         if (newG.fastEquals(generator)) {
           g
         } else {
           Generate(newG.asInstanceOf[Generator], join, outer, qualifier, output, child)
         }
 
+      case o: ObjectOperator if containsUnresolvedDeserializer(o.deserializers.map(_._1)) =>
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Add a comment about why this case is special.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala (hunk 2 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala]: No they are different. Like `MapGroups`, the child.output also contains grouping attributes, which may cause problems if we have same name attribute in grouping attributes and data attributes.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/object.scala (hunk 4 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/object.scala]: according to the definition of serializer: `def serializer: Seq[NamedExpression]`, a seq of named expressions work together as a serialize, so `withNewSerializer` seems more accuracy here?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala (hunk 2 above)
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala]: I might be missing something here, but it looks like there is only one place where we create a `CoGroup` operator and it is always using just the left/right output attributes.  Lets not over generalize unless it is useful.

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala]: yea there is only one place where we create a `CoGroup`, but we do pass `GroupedDataset.dataAttributes` instead of left/right output attributes, https://github.com/apache/spark/pull/10852/files#diff-ed737c05b572cbec98ae5264d2c799b4R314

Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala]: I see, thanks for explaining.

**Comment by marmbrus** [comment]
I'm sorry... I let this go stale.

**Comment by cloud-fan** [comment]
oh, didn't notice it. Will update it soon

**Comment by marmbrus** [comment]
Thanks, merging to master.
